dm/dm/master/openapi_controller.go

```diff
@@ -0,0 +1,288 @@
+// Copyright 2022 PingCAP, Inc.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// See the License for the specific language governing permissions and
+// limitations under the License.
+//
+// MVC for dm-master's openapi server
+// Model(data in etcd): source of truth
+// View(openapi_view): do some inner work such as validate, filter, prepare parameters/response and call controller to update model.
+// Controller(openapi_controller): call model func to update data.
+
+package master
+
+import (
+	"context"
+
+	"github.com/pingcap/tiflow/dm/checker"
+	"github.com/pingcap/tiflow/dm/dm/config"
+	"github.com/pingcap/tiflow/dm/dm/pb"
+	"github.com/pingcap/tiflow/dm/openapi"
+	"github.com/pingcap/tiflow/dm/pkg/terror"
+)
+
+func (s *Server) getSourceStatusListFromWorker(ctx context.Context, sourceName string, specifiedSource bool) ([]openapi.SourceStatus, error) {
+	workerStatusList := s.getStatusFromWorkers(ctx, []string{sourceName}, "", specifiedSource)
+	sourceStatusList := make([]openapi.SourceStatus, len(workerStatusList))
+	for i, workerStatus := range workerStatusList {
+		if workerStatus == nil {
+			// this should not happen unless the rpc in the worker server has been modified
+			return nil, terror.ErrOpenAPICommonError.New("worker's query-status response is nil")
+		}
+		sourceStatus := openapi.SourceStatus{SourceName: sourceName, WorkerName: workerStatus.SourceStatus.Worker}
+		if !workerStatus.Result {
+			sourceStatus.ErrorMsg = &workerStatus.Msg
+		} else if relayStatus := workerStatus.SourceStatus.GetRelayStatus(); relayStatus != nil {
+			sourceStatus.RelayStatus = &openapi.RelayStatus{
+				MasterBinlog:       relayStatus.MasterBinlog,
+				MasterBinlogGtid:   relayStatus.MasterBinlogGtid,
+				RelayBinlogGtid:    relayStatus.RelayBinlogGtid,
+				RelayCatchUpMaster: relayStatus.RelayCatchUpMaster,
+				RelayDir:           relayStatus.RelaySubDir,
+				Stage:              relayStatus.Stage.String(),
+			}
+		}
+		sourceStatusList[i] = sourceStatus
+	}
+	return sourceStatusList, nil
+}
+
+// nolint:unparam
+func (s *Server) createSource(ctx context.Context, cfg *config.SourceConfig) error {
```

Comment:
seprate test for this file we be added in next pr, current all func is tested Indirectly over `openapi_view_test.go`